File: b-tree.cs
/*
    Tomas Petrzela
    INF - UPOL
    mailto:tomas.petrzela02@upol.cz
    _______________________________
    ALGO2 - Zapoctova prace - B strom
*/
using System;
using System.Diagnostics;

var sw = new Stopwatch();
sw.Start();

var tree = new Btree(3);

var insertArray = new[]
    { 5, 7, 15, 40, 2, 1, 3, 10, 12, 80, 200, 500, 150, 90, 20, 50, 16, 14, 13, 9, 38, 89, 101, 102, 201, 202, 91, 88, 30, 18 };
foreach (var elem in insertArray)
    tree.Insert(elem);

Console.WriteLine("Search(5): {0}", tree.Search(5));
Console.WriteLine("Search(999): {0}", tree.Search(999));
Console.WriteLine("FindMaxKey(): {0}", tree.FindMaxKey());

Console.WriteLine("Delete(5): {0}", tree.Delete(5));
Console.WriteLine("Search(5): {0}", tree.Search(5));

var deleteArray = new[]
    { 15, 1, 80, 50, 14, 12, 500, 89, 90, 30, 150, 7, 13, 10, 16, 201, 102 };
foreach (var elem in deleteArray)
    tree.Delete(elem);

Console.WriteLine("FindMaxKey(): {0}", tree.FindMaxKey());

tree.Print();

sw.Stop();
Console.WriteLine("Elapsed: {0}", sw.Elapsed);

class BtreeNode
{
    public BtreeNode Parent;
    private int T { get; }
    public int Length;
    public readonly int?[] Keys;
    public readonly BtreeNode[] Childrens;
    public readonly bool IsLeaf;

    public BtreeNode(int t)
    {
        T = t;
        Length = 0;
        Keys = new int?[2 * t - 1];
        Childrens = new BtreeNode[2 * t];

        IsLeaf = false;
    }

    public BtreeNode(int t, bool leaf)
    {
        T = t;
        Length = 0;
        Keys = new int?[2 * t - 1];
        Childrens = new BtreeNode[2 * t];

        IsLeaf = leaf;
    }

    public int? GetFirstKey() => Length == 0 ? null : Keys[0];

    public int? GetFirstKeyIndex() => Length == 0 ? null : 0;

    public int? GetLastKey() => Length == 0 ? null : Keys[Length - 1];

    public int? GetLastKeyIndex() => Length == 0 ? null : Length - 1;

    public int? GetLastChildrenIndex()
    {
        for (var i = Childrens.Length - 1; i >= 0; i--)
            if (Childrens[i] != null)
                return i;

        return null;
    }

    public BtreeNode GetLastChildren()
    {
        for (var i = Childrens.Length - 1; i >= 0; i--)
            if (Childrens[i] != null)
                return Childrens[i];

        return null;
    }

    public int?[] GetHashCodeOfChildrens()
    {
        var result = new int?[2 * T];

        for (var i = 0; i < 2 * T; i++)
            if (Childrens[i] != null)
                result[i] = Childrens[i].GetHashCode();

        return result;
    }

    public (int? key, int? index) FindMaxKey()
    {
        if (IsLeaf)
            return (GetLastKey(), GetLastKeyIndex());
        else
        {
            return GetLastChildren().FindMaxKey();
        }
    }

    public void InsertInOrder(int key)
    {
        if (Length == 0)
            Keys[0] = key;
        else
        {
            var i = Length - 1;
            while (i >= 0 && key < Keys[i])
            {
                Keys[i + 1] = Keys[i];
                i--;
            }

            Keys[i + 1] = key;
        }

        Length++;
    }

    public void InsertNonFull(int key)
    {
        if (IsLeaf)
            InsertInOrder(key);
        else
        {
            var i = Length - 1;

            while (i >= 0 && Keys[i] > key)
                i--;

            i++;

            if (Childrens[i].Length == 2 * T - 1)
            {
                SplitChild(i);

                if (key > Keys[i])
                    i++;
            }

            Childrens[i].InsertNonFull(key);
        }
    }

    public void SplitChild(int index)
    {
        var y = Childrens[index];
        var z = new BtreeNode(T, y.IsLeaf)
        {
            Parent = y.Parent,
            Length = T - 1
        };

        y.Length = T - 1;

        // 1.
        // Zkopírování pravé částí klíčů do nového uzlu
        for (var j = 0; j < T - 1; j++)
        {
            z.Keys[j] = y.Keys[j + T];
            y.Keys[j + T] = null;
        }

        if (!y.IsLeaf)
            // Zkopírování pravé částí potomků do nového uzlu
            for (var j = 0; j < T; j++)
            {
                z.Childrens[j] = y.Childrens[j + T];
                y.Childrens[j + T] = null;

                // Potomci v novém uzlu budou mít správného rodiče
                z.Childrens[j].Parent = z;
            }

        // 2.
        // Posuneme klíče v aktuálním uzlu o 1 doprava
        for (var j = Length - 1; j >= index; j--)
            Keys[j + 1] = Keys[j];

        // Posuneme potomky v aktuálním uzlu o 1 doprava
        for (var j = Length; j >= index + 1; j--)
            Childrens[j + 1] = Childrens[j];

        // 3.
        // Do akzuálního uzlo zapojíme nový uzel
        Childrens[index + 1] = z;
        Keys[index] = y.Keys[T - 1];
        y.Keys[T - 1] = null;
        Length++;
    }

    public (BtreeNode node, int index) Search(int key)
    {
        var i = 0;
        while (i < Length && key > Keys[i])
            i++;

        if (i < Length && key == Keys[i])
            return (this, i);

        if (IsLeaf)
            return (null, 0);

        return Childrens[i].Search(key);
    }

    private void MargeTwoNodes(bool leftSibling, int thisNodeIndex)
    {
        var mergedNode = new BtreeNode(T, IsLeaf) { Parent = this.Parent };
        BtreeNode siblingNode;

        if (!leftSibling)
        {
            InsertThisNode();

            InsertParentKey();

            InsertSibling();

            Parent.Childrens[thisNodeIndex] = mergedNode;
            MoveParentChildrens(thisNodeIndex + 1);
        }
        else
        {
            InsertSibling();

            InsertParentKey();

            InsertThisNode();

            Parent.Childrens[thisNodeIndex - 1] = mergedNode;
            MoveParentChildrens(thisNodeIndex);
        }

        void InsertThisNode()
        {
            var length = mergedNode.Length;
            for (var i = 0; i < Length; i++)
            {
                mergedNode.Keys[length + i] = Keys[i];
                Keys[i] = null;
                mergedNode.Length++;
            }

            for (var i = 0; i < GetLastChildrenIndex(); i++)
            {
                mergedNode.Childrens[i] = Childrens[i];
                Childrens[i] = null;
            }
        }

        void InsertSibling()
        {
            siblingNode = !leftSibling ? Parent.Childrens[thisNodeIndex + 1] : Parent.Childrens[thisNodeIndex - 1];

            var length = mergedNode.Length;
            for (var i = 0; i < siblingNode.Length; i++)
            {
                mergedNode.Keys[length + i] = siblingNode.Keys[i];
                siblingNode.Keys[i] = null;
                mergedNode.Length++;
            }

            for (var i = 0; i < siblingNode.GetLastChildrenIndex(); i++)
            {
                mergedNode.Childrens[length + i] = siblingNode.Childrens[i];
                siblingNode.Childrens[i] = null;
            }
        }

        void InsertParentKey()
        {
            int index;
            if (!leftSibling)
                index = thisNodeIndex;
            else
                index = thisNodeIndex - 1;

            mergedNode.Keys[mergedNode.Length] = Parent.Keys[index];

            for (var i = index; i < 2 * T - 1; i++)
            {
                if (Parent.Keys[i] == null)
                    break;

                Parent.Keys[i] = Parent.Keys[i + 1];
            }

            mergedNode.Length++;
        }

        void MoveParentChildrens(int startIndex)
        {
            for (var i = startIndex; i < 2 * T - 1; i++)
                Parent.Childrens[i] = Parent.Childrens[i + 1];
        }
    }

    public void Delete(int index)
    {
        DeletePhaseOne(index);

        DeletePhaseTwo();
    }

    private int? DeletePhaseOne(int index)
    {
        return IsLeaf ? DeleteFromLeaf(index) : DeleteFromNoLeaf(index);
    }

    private int? DeleteFromLeaf(int index)
    {
        var temp = Keys[index];

        for (var i = index; i < Length - 1; i++)
            Keys[i] = Keys[i + 1];

        Length--;

        for (var i = Length; i < 2 * T - 1; i++)
        {
            if (Keys[i] == null)
                break;

            Keys[i] = null;
        }

        return temp;
    }

    private int? DeleteFromNoLeaf(int index)
    {
        var temp = Keys[index];

        var (key, _) = Childrens[index].FindMaxKey();
        var (node, i) = Search(key ?? default(int));
        Keys[index] = key;

        node.Delete(i);

        return temp;
    }

    private void DeletePhaseTwo()
    {
        // 1.
        if (Parent == null
            || Length >= T - 1)
            return;

        var thisNodeIndex = Array.IndexOf(Parent.Childrens, this);

        // 2.
        // Levý sourozenec
        if (thisNodeIndex > 0
            && Parent.Childrens[thisNodeIndex - 1].Length > T - 1)
        {
            MoveKeyOverParent(true);
            return;
        }
        // Pravý sourozenec
        else if (thisNodeIndex < 2 * T
          && Parent.Childrens[thisNodeIndex + 1] != null
          && Parent.Childrens[thisNodeIndex + 1].Length > T - 1)
        {
            MoveKeyOverParent(false);
            return;
        }

        // 3.
        // Pravý sourozenec
        if (thisNodeIndex >= 0 && thisNodeIndex < 2 * T
            && Parent.Childrens[thisNodeIndex + 1] != null
            && Parent.Childrens[thisNodeIndex + 1].Length >= T - 1)
        {
            MargeTwoNodes(false, thisNodeIndex);
            Parent.DeletePhaseTwo();
        }
        // Levý sourozenec
        else if (thisNodeIndex > 0 && thisNodeIndex <= 2 * T
            && Parent.Childrens[thisNodeIndex - 1].Length >= T - 1)
        {
            MargeTwoNodes(true, thisNodeIndex);
            Parent.DeletePhaseTwo();
        }

        void MoveKeyOverParent(bool leftSibling)
        {
            int siblingIndex;
            int siblingKeyIndex;
            int keyIndex;

            if (leftSibling)
            {
                siblingIndex = thisNodeIndex - 1;
                siblingKeyIndex = Parent.Childrens[siblingIndex].GetLastKeyIndex() ?? default(int);
                keyIndex = thisNodeIndex - 1;
            }
            else
            {
                siblingIndex = thisNodeIndex + 1;
                siblingKeyIndex = Parent.Childrens[siblingIndex].GetFirstKeyIndex() ?? default(int);
                keyIndex = thisNodeIndex;
            }

            InsertInOrder(Parent.Keys[keyIndex] ?? default(int));

            for (var i = GetLastChildrenIndex(); i < 2 * T - 1; i++)
                Childrens[(int)i] = Childrens[(int)i];

            // Přelití posledního potomka sourozence
            Childrens[0] = Parent.Childrens[siblingIndex].GetLastChildren();
            Parent.Childrens[siblingIndex].Childrens[GetLastChildrenIndex() ?? default(int)] = null;

            Parent.Keys[keyIndex] =
                Parent.Childrens[siblingIndex].DeletePhaseOne(siblingKeyIndex);
        }
    }

    public void Print()
    {
        Console.WriteLine("this: {0}", GetHashCode());
        if (Parent != null)
            Console.WriteLine("parent: {0}", Parent.GetHashCode());
        Console.WriteLine("keys: [{0}]", String.Join(", ", Keys));
        Console.WriteLine("childrens: [{0}]", String.Join(", ", GetHashCodeOfChildrens()));
        Console.WriteLine("leaf: {0}", IsLeaf);
        Console.WriteLine();

        foreach (var children in Childrens)
            children?.Print();
    }

    public void PrintRoot()
    {
        if (Parent == null)
            Print();
        else
            Parent.PrintRoot();
    }
}

class Btree
{
    public int T { get; }
    public BtreeNode Root;

    public Btree(int t)
    {
        if (t <= 2)
            throw new ArgumentException("BTree");

        T = t;
        Root = new BtreeNode(t, true);
    }

    public void Insert(int key)
    {
        var root = Root;

        if (root.Length == 0)
            root.InsertInOrder(key);
        else if (root.Length == 2 * T - 1)
        {
            var newNode = new BtreeNode(T);
            newNode.Childrens[0] = root;
            Root = newNode;
            root.Parent = newNode;

            newNode.SplitChild(0);
            newNode.InsertNonFull(key);
        }
        else
            root.InsertNonFull(key);
    }

    public (BtreeNode node, int index) Search(int key) => Root.Search(key);

    public (int? key, int? index) FindMaxKey() => Root.FindMaxKey();

    public int? Delete(int key)
    {
        // Tree is empty
        if (Root.Length == 0)
            return null;

        var (node, index) = Search(key);

        // Node not found
        if (node == null)
            return null;

        node.Delete(index);

        return key;
    }

    public void Print()
    {
        Console.WriteLine("________________");
        Console.WriteLine("Btree");
        Console.WriteLine("t: {0}", T);

        if (Root == null) return;
        Console.WriteLine("\n________________");
        Console.WriteLine("R O O T:");
        Root.Print();
    }
}

File: b-tree.go
package main

import (
	"errors"
	"reflect"
)

type node struct {
	parent    *node   // rodič
	T         int     //
	length    int     //
	keys      []int   // klíč
	childrens []*node // potomci
	isLeaf    bool    // jsem listem
}

func NewBtreeNode(t int, isLeaf bool) *node {
	return &node{
		T:         t,
		length:    0,
		keys:      make([]int, 2*t-1),
		childrens: make([]*node, 2*t),
		isLeaf:    isLeaf,
	}
}

func (this node) GetFirstKey() int {
	if this.length == 0 {
		return -1
	} else {
		return this.keys[0]
	}
}

func (this node) GetFirstKeyIndex() int {

	if this.length == 0 {
		return -1
	} else {
		return 0
	}
}

func (this node) GetLastKey() int {
	if this.length == 0 {
		return -1
	} else {
		return this.keys[this.length-1]
	}
}

func (this node) GetLastKeyIndex() int {

	if this.length == 0 {
		return -1
	} else {
		return this.length - 1
	}
}

func (this node) GetLastChildrenIndex() int {
	for i := len(this.childrens) - 1; i >= 0; i-- {
		if this.childrens[i] != nil {
			return i
		}
	}

	return -1
}

func (this node) GetLastChildren() *node {
	for i := len(this.childrens) - 1; i >= 0; i-- {
		if this.childrens[i] != nil {
			return this.childrens[i]
		}
	}

	return nil
}

func (this node) FindMaxKey() (int, int) {
	if this.isLeaf {
		return this.GetLastKey(), this.GetFirstKeyIndex()
	} else {
		return this.GetLastChildren().FindMaxKey()
	}
}

func (this node) InsertInOrder(key int) {
	if this.length == 0 {
		this.keys[0] = key
	} else {
		i := this.length - 1

		for i >= 0 && key < this.keys[i] {
			this.keys[i+1] = this.keys[i]
			i--
		}

		this.keys[i+1] = key
	}

	this.length++
}

func (this node) InsertNonFull(key int) {
	if this.isLeaf {
		this.InsertInOrder(key)
	} else {
		i := this.length - 1

		for i >= 0 && this.keys[i] > key {
			i--
		}

		i++

		if this.childrens[i].length == 2*this.T-1 {
			this.SplitChild(i)

			if key > this.keys[i] {
				i++
			}
		}

		this.childrens[i].InsertNonFull(key)
	}
}

func (this node) SplitChild(index int) {
	y := this.childrens[index]
	z := NewBtreeNode(this.T, y.isLeaf)

	z.parent = y.parent
	z.length = this.T - 1
	y.length = this.T - 1

	// 1.
	// Zkopírování pravé částí klíčů do nového uzlu
	for j := 0; j < this.T; j++ {
		z.keys[j] = y.keys[j+this.T]
		y.keys[j+this.T] = 0
	}

	if !y.isLeaf {
		// Zkopírování pravé částí potomků do nového uzlu
		for j := 0; j < this.T; j++ {
			z.childrens[j] = y.childrens[j+this.T]
			y.childrens[j+this.T] = nil

			// Potomci v novém uzlu budou mít správného rodiče
			z.childrens[j].parent = z
		}
	}

	// 2.
	// Posuneme klíče v aktuálním uzlu o 1 doprava

	for j := this.length - 1; j >= index; j-- {
		this.keys[j+1] = this.keys[j]
	}

	// Posuneme potomky v aktuálním uzlu o 1 doprava
	for j := this.length; j >= index+1; j-- {
		this.childrens[j+1] = this.childrens[j]
	}

	// 3.
	// Do akzuálního uzlo zapojíme nový uzel
	this.childrens[index+1] = z
	this.keys[index] = y.keys[this.T-1]
	y.keys[this.T-1] = 0
	this.length++
}

func (this node) Search(key int) (*node, int) {
	i := 0

	for i < this.length && key > this.keys[i] {
		i++
	}

	if i < this.length && key == this.keys[i] {
		return &this, i
	}

	if this.isLeaf {
		return nil, -1
	}

	return this.childrens[i].Search(key)
}

func (this node) InsertThisNode(mergedNode *node) {
	length := mergedNode.length
	for i := 0; i < this.length; i++ {
		mergedNode.keys[length+i] = this.keys[i]
		this.keys[i] = 0
		mergedNode.length++
	}

	for i := 0; i < this.GetLastChildrenIndex(); i++ {
		mergedNode.childrens[i] = this.childrens[i]
		this.childrens[i] = nil
	}
}

func (this node) InsertSiblig(mergedNode *node, leftSibling bool, thisNodeIndex int) {
	var siblingNode *node
	if !leftSibling {
		siblingNode = this.parent.childrens[thisNodeIndex+1]
	} else {
		siblingNode = this.parent.childrens[thisNodeIndex-1]
	}

	length := mergedNode.length
	for i := 0; i < siblingNode.length; i++ {
		mergedNode.keys[length+i] = siblingNode.keys[i]
		siblingNode.keys[i] = 0
		mergedNode.length++
	}

	for i := 0; i < siblingNode.GetLastChildrenIndex(); i++ {
		mergedNode.childrens[length+i] = siblingNode.childrens[i]
		siblingNode.childrens[i] = nil
	}
}

func (this node) InsertParentKey(mergedNode *node, leftSibling bool, thisNodeIndex int) {
	var index int

	if !leftSibling {
		index = thisNodeIndex
	} else {
		index = thisNodeIndex - 1
	}

	mergedNode.keys[mergedNode.length] = this.parent.keys[index]

	for i := index; i < 2*this.T-1; i++ {
		if this.parent.keys[i] == 0 {
			break
		}

		this.parent.keys[i] = this.parent.keys[i+1]
	}

	mergedNode.length++
}

func (this node) MoveParentChildrens(startIndex int) {
	for i := startIndex; i < 2*this.T-1; i++ {
		this.parent.childrens[i] = this.parent.childrens[i+1]
	}
}

func (this node) MargeTwoNodes(leftSibling bool, thisNodeIndex int) {
	mergedNode := NewBtreeNode(this.T, this.isLeaf)
	mergedNode.parent = this.parent

	if !leftSibling {
		this.InsertThisNode(mergedNode)
		this.InsertParentKey(mergedNode, leftSibling, thisNodeIndex)
		this.InsertSiblig(mergedNode, leftSibling, thisNodeIndex)

		this.parent.childrens[thisNodeIndex] = mergedNode
		this.MoveParentChildrens(thisNodeIndex + 1)
	} else {
		this.InsertSiblig(mergedNode, leftSibling, thisNodeIndex)
		this.InsertParentKey(mergedNode, leftSibling, thisNodeIndex)
		this.InsertThisNode(mergedNode)

		this.parent.childrens[thisNodeIndex-1] = mergedNode
		this.MoveParentChildrens(thisNodeIndex)
	}
}

func (this node) Delete(index int) {
	this.DeletePhaseOne(index)
	this.DeletePhaseTwo()
}

func (this node) DeletePhaseOne(index int) int {
	if this.isLeaf {
		return this.DeleteFromLeaf(index)
	} else {
		return this.DeleteFromNoLeaf(index)
	}
}

func (this node) DeleteFromLeaf(index int) int {
	temp := this.keys[index]

	for i := index; i < this.length-1; i++ {
		this.keys[i] = this.keys[i+1]
	}

	this.length--

	for i := this.length; i < 2*this.T-1; i++ {
		if this.keys[i] == 0 {
			break
		}

		this.keys[i] = 0
	}

	return temp
}

func (this node) DeleteFromNoLeaf(index int) int {
	temp := this.keys[index]

	key, _ := this.childrens[index].FindMaxKey()
	node, i := this.Search(key)
	this.keys[index] = key

	node.Delete(i)

	return temp
}

func (this node) MoveKeyOverParent(leftSibling bool, thisNodeIndex int) {
	var siblingIndex int
	var siblingKeyIndex int
	var keyIndex int

	if leftSibling {
		siblingIndex = thisNodeIndex - 1
		siblingKeyIndex = this.parent.childrens[siblingIndex].GetLastChildrenIndex()
		keyIndex = thisNodeIndex - 1
	} else {
		siblingIndex = thisNodeIndex + 1
		siblingKeyIndex = this.parent.childrens[siblingIndex].GetFirstKeyIndex()
		keyIndex = thisNodeIndex
	}

	this.InsertInOrder(this.parent.keys[keyIndex])

	for i := this.GetLastChildrenIndex(); i < 2*this.T-1; i++ {
		/* this.childrens[i] = this.childrens[i] */
	}

	// Přelití posledního potomka sourozence
	this.childrens[0] = this.parent.childrens[siblingIndex].GetLastChildren()
	this.parent.childrens[siblingIndex].childrens[this.GetLastChildrenIndex()] = nil

	this.parent.keys[keyIndex] = this.parent.childrens[siblingIndex].DeletePhaseOne(siblingKeyIndex)
}

func (this node) IndexOf() int {
	for i := 0; i < len(this.childrens); i++ {
		if reflect.DeepEqual(this.childrens[i], this) {
			return i
		}
	}

	return -1
}

func (this node) DeletePhaseTwo() {
	// 1.
	if this.parent == nil || this.length >= this.T-1 {
		return
	}

	thisNodeIndex := this.IndexOf()

	// 2.
	// Levý sourozenec
	if thisNodeIndex > 0 && this.parent.childrens[thisNodeIndex-1].length > this.T-1 {
		this.MoveKeyOverParent(true, thisNodeIndex)
		return

		// Pravý sourozenec
	} else if thisNodeIndex < 2*this.T && this.parent.childrens[thisNodeIndex+1] != nil && this.parent.childrens[thisNodeIndex+1].length > this.T-1 {
		this.MoveKeyOverParent(false, thisNodeIndex)
		return
	}

	// 3.
	// Pravý sourozenec
	if thisNodeIndex >= 0 && thisNodeIndex < 2*this.T && this.parent.childrens[thisNodeIndex+1] != nil && this.parent.childrens[thisNodeIndex+1].length >= this.T-1 {
		this.MargeTwoNodes(false, thisNodeIndex)
		this.parent.DeletePhaseTwo()

		// Levý sourozenec
	} else if thisNodeIndex > 0 && thisNodeIndex <= 2*this.T && this.parent.childrens[thisNodeIndex-1].length >= this.T-1 {
		this.MargeTwoNodes(true, thisNodeIndex)
		this.parent.DeletePhaseTwo()
	}
}

type tree struct {
	T    int   //
	root *node // kořenový uzel
}

func NewBtree(t int) tree {
	if t < 2 {
		panic(errors.New("t cannot be less than 2"))
	}

	return tree{
		T:    t,
		root: NewBtreeNode(t, true),
	}
}

func (this tree) Insert(key int) {
	root := this.root

	if root.length == 0 {
		root.InsertInOrder(key)
	} else if root.length == 2*this.T-1 {
		var newNode = NewBtreeNode(this.T, false)
		newNode.childrens[0] = root
		this.root = newNode
		root.parent = newNode

		newNode.SplitChild(0)
		newNode.InsertNonFull(key)
	} else {
		root.InsertNonFull(key)
	}
}

func (this tree) Search(key int) (*node, int) {
	return this.root.Search(key)
}

func (this tree) FindMaxKey() (int, int) {
	return this.root.FindMaxKey()
}

func (this tree) Delete(key int) int {
	// Tree is empty
	if this.root.length == 0 {
		return -1
	}

	node, index := this.Search(key)

	// Node not found
	if node == nil {
		return -1
	}

	node.Delete(index)

	return key
}
